hbase-server/src/main/java/org/apache/hadoop/hbase/master/MetaFixer.java

```diff
@@ -86,74 +87,176 @@ void fix() throws IOException {
    * If hole, it papers it over by adding a region in the filesystem and to hbase:meta.
    * Does not assign.
    */
-  void fixHoles(CatalogJanitor.Report report) throws IOException {
-    List<Pair<RegionInfo, RegionInfo>> holes = report.getHoles();
+  void fixHoles(CatalogJanitor.Report report) {
+    final List<Pair<RegionInfo, RegionInfo>> holes = report.getHoles();
     if (holes.isEmpty()) {
-      LOG.debug("No holes.");
+      LOG.info("CatalogJanitor Report contains no holes to fix. Skipping.");
       return;
     }
-    for (Pair<RegionInfo, RegionInfo> p: holes) {
-      RegionInfo ri = getHoleCover(p);
-      if (ri == null) {
-        continue;
-      }
-      Configuration configuration = this.masterServices.getConfiguration();
-      HRegion.createRegionDir(configuration, ri, FSUtils.getRootDir(configuration));
-      // If an error here, then we'll have a region in the filesystem but not
-      // in hbase:meta (if the below fails). Should be able to rerun the fix.
-      // Add to hbase:meta and then update in-memory state so it knows of new
-      // Region; addRegionToMeta adds region and adds a state column set to CLOSED.
-      MetaTableAccessor.addRegionToMeta(this.masterServices.getConnection(), ri);
-      this.masterServices.getAssignmentManager().getRegionStates().
-          updateRegionState(ri, RegionState.State.CLOSED);
-      LOG.info("Fixed hole by adding {} in CLOSED state; region NOT assigned (assign to ONLINE).",
-          ri);
-    }
+
+    LOG.info("Identified {} region holes to fix.", holes.size());
+
+    final List<RegionInfo> newRegionInfos = mapHolesToRegionInfos(holes);
+    final List<RegionInfo> newRegionDirectories =
+      createRegionDirectories(masterServices, newRegionInfos);
+    final List<RegionInfo> newMetaEntries = createMetaEntries(masterServices, newRegionDirectories);
+    final List<TransitRegionStateProcedure> scheduledProcedures =
+      masterServices.getAssignmentManager().createAssignProceduresBestEffort(newMetaEntries);
+    LOG.info(
+      "Scheduled {}/{} new regions for assignment.", scheduledProcedures.size(), holes.size());
   }
 
   /**
-   * @return Calculated RegionInfo that covers the hole <code>hole</code>
+   * Create a new {@link RegionInfo} corresponding to each provided "hole" pair.
    */
-  private RegionInfo getHoleCover(Pair<RegionInfo, RegionInfo> hole) {
-    RegionInfo holeCover = null;
-    RegionInfo left = hole.getFirst();
-    RegionInfo right = hole.getSecond();
+  private static List<RegionInfo> mapHolesToRegionInfos(
```

Comment:
Good making it static.

We are creating RIs in here, no? Not 'map'ping holes to RIs? Gives impression the RIs were there already rather than being created.